File: exam-slammr-webapp/src/main/java/uk/co/claritysoftware/exam/slammr/webapp/web/controller/QuestionController.java
package uk.co.claritysoftware.exam.slammr.webapp.web.controller;

import static uk.co.claritysoftware.exam.slammr.webapp.web.factory.CreateQuestionFactory.valueOf;
import static uk.co.claritysoftware.exam.slammr.webapp.web.model.question.AnswerOption.generateEmptyAnswerOption;
import static uk.co.claritysoftware.exam.slammr.webapp.web.model.question.CreateQuestion.generateEmptyCreateQuestion;
import static uk.co.claritysoftware.exam.slammr.webapp.web.model.question.FurtherReading.generateEmptyFurtherReading;

import java.security.Principal;
import java.util.Optional;
import javax.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.servlet.ModelAndView;
import uk.co.claritysoftware.exam.slammr.webapp.service.QuestionService;
import uk.co.claritysoftware.exam.slammr.webapp.service.model.question.Question;
import uk.co.claritysoftware.exam.slammr.webapp.web.exception.QuestionNotFoundException;
import uk.co.claritysoftware.exam.slammr.webapp.web.model.question.CreateQuestion;

import lombok.extern.slf4j.Slf4j;

/**
 * Controller to handle question page requests
 */
@Slf4j
@Controller
@RequestMapping("/question")
public class QuestionController {

    private final QuestionService questionService;

    @Autowired
    public QuestionController(QuestionService questionService) {
        this.questionService = questionService;
    }

    @GetMapping("/new")
    public ModelAndView getQuestionPage() {
        return createQuestionViewHidingBindErrors(generateEmptyCreateQuestion());
    }

    @PostMapping
    public ModelAndView createNewQuestion(@Valid @ModelAttribute("form") CreateQuestion form, BindingResult bindingResult, Principal principal) {

        switch (form.getAction()) {
            case addCertification:
                form.getCertifications().add("");
                return createQuestionViewHidingBindErrors(form);

            case addTag:
                form.getTags().add("");
                return createQuestionViewHidingBindErrors(form);

            case addFurtherReading:
                form.getFurtherReadings().add(generateEmptyFurtherReading());
                return createQuestionViewHidingBindErrors(form);

            case addAnswer:
                form.getAnswerOptions().add(generateEmptyAnswerOption());
                return createQuestionViewHidingBindErrors(form);

            default:
                if (bindingResult.hasErrors()) {
                    return createQuestionViewShowingBindErrors(form);
                }

                Question newQuestion = valueOf(form, principal.getName());
                Question savedQuestion = questionService.saveNewQuestion(newQuestion);

                return new ModelAndView("redirect:/question/" + savedQuestion.getId() + "/" + savedQuestion.getSlug());
        }
    }

    @GetMapping("/{id}/{slug}")
	public ModelAndView getExistingQuestionPage(@PathVariable("id") String id, @PathVariable("slug") String slug) {
    	Optional<Question> retrievedQuestion = questionService.getQuestionById(id);
    	if (!retrievedQuestion.isPresent()) {
    		log.warn("Question with id {} could not be found", id);
    		throw new QuestionNotFoundException(id);
		}

		Question question = retrievedQuestion.get();

		if (!question.getSlug().equals(slug)) {
			String redirectUrl = "redirect:/question/" + id + "/" + question.getSlug();
			log.warn("Question with id {} does not have requested slug {}. Redirecting to {}", id, slug, redirectUrl);
			return new ModelAndView(redirectUrl);
		}

		ModelAndView modelAndView = new ModelAndView("question/view");
		modelAndView.addObject("question", question);
    	return modelAndView;
	}

    @GetMapping("/{id}")
	public ModelAndView getExistingQuestionPageWithoutSlug(@PathVariable("id") String id) {
    	return getExistingQuestionPage(id, null);
	}

    private ModelAndView createQuestionViewShowingBindErrors(CreateQuestion form) {
        ModelAndView modelAndView = new ModelAndView("question/create");
        modelAndView.addObject("showBindErrors", true);
		modelAndView.addObject("form", form);
        return modelAndView;
    }

    private ModelAndView createQuestionViewHidingBindErrors(CreateQuestion form) {
        ModelAndView modelAndView = new ModelAndView("question/create");
        modelAndView.addObject("showBindErrors", false);
        modelAndView.addObject("form", form);
        return modelAndView;
    }
}

File: exam-slammr-webapp/src/main/kotlin/uk/co/claritysoftware/exam/slammr/webapp/web/controller/QuestionController.kt
package uk.co.claritysoftware.exam.slammr.webapp.web.controller

import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Controller
import org.springframework.validation.BindingResult
import org.springframework.web.bind.annotation.*
import org.springframework.web.servlet.ModelAndView
import uk.co.claritysoftware.exam.slammr.webapp.service.QuestionService
import uk.co.claritysoftware.exam.slammr.webapp.web.exception.QuestionNotFoundException
import uk.co.claritysoftware.exam.slammr.webapp.web.factory.CreateQuestionFactory.Companion.valueOf
import uk.co.claritysoftware.exam.slammr.webapp.web.model.question.AnswerOption
import uk.co.claritysoftware.exam.slammr.webapp.web.model.question.CreateQuestion
import uk.co.claritysoftware.exam.slammr.webapp.web.model.question.FurtherReading
import java.security.Principal
import javax.validation.Valid

/**
 * Controller to handle question page requests
 */
@Controller
@RequestMapping("/question")
class QuestionController(@Autowired val questionService: QuestionService) {

	private companion object {
		val log = LoggerFactory.getLogger(QuestionController::class.java)
	}

	@GetMapping("/new")
	fun getQuestionPage(): ModelAndView {
		return createQuestionViewHidingBindErrors(CreateQuestion())
	}

	@PostMapping
	fun createNewQuestion(@Valid @ModelAttribute("form") form: CreateQuestion, bindingResult: BindingResult, principal: Principal): ModelAndView {

		when (form.action) {
			CreateQuestion.Action.addCertification -> {
				form.certifications.add("")
				return createQuestionViewHidingBindErrors(form)
			}

			CreateQuestion.Action.addTag -> {
				form.tags.add("")
				return createQuestionViewHidingBindErrors(form)
			}

			CreateQuestion.Action.addFurtherReading -> {
				form.furtherReadings.add(FurtherReading())
				return createQuestionViewHidingBindErrors(form)
			}

			CreateQuestion.Action.addAnswer -> {
				form.answerOptions.add(AnswerOption())
				return createQuestionViewHidingBindErrors(form)
			}

			else -> {

				if (bindingResult.hasErrors()) {
					return createQuestionViewShowingBindErrors(form)
				}

				val newQuestion = valueOf(form, principal.name)
				val (id, slug) = questionService.saveNewQuestion(newQuestion)

				return ModelAndView("redirect:/question/$id/$slug")
			}
		}
	}

	@GetMapping("/{id}/{slug}")
	fun getExistingQuestionPage(@PathVariable("id") id: String, @PathVariable("slug") slug: String?): ModelAndView {
		val retrievedQuestion = questionService.getQuestionById(id)
		if (!retrievedQuestion.isPresent) {
			log.warn("Question with id {} could not be found", id)
			throw QuestionNotFoundException(id)
		}

		val question = retrievedQuestion.get()

		if (question.slug != slug) {
			val redirectUrl = "redirect:/question/" + id + "/" + question.slug
			log.warn("Question with id {} does not have requested slug {}. Redirecting to {}", id, slug, redirectUrl)
			return ModelAndView(redirectUrl)
		}

		val modelAndView = ModelAndView("question/view")
		modelAndView.addObject("question", question)
		return modelAndView
	}

	@GetMapping("/{id}")
	fun getExistingQuestionPageWithoutSlug(@PathVariable("id") id: String): ModelAndView {
		return getExistingQuestionPage(id, null)
	}

	private fun createQuestionViewShowingBindErrors(form: CreateQuestion): ModelAndView {
		val modelAndView = ModelAndView("question/create")
		modelAndView.addObject("showBindErrors", true)
		modelAndView.addObject("form", form)
		return modelAndView
	}

	private fun createQuestionViewHidingBindErrors(form: CreateQuestion): ModelAndView {
		val modelAndView = ModelAndView("question/create")
		modelAndView.addObject("showBindErrors", false)
		modelAndView.addObject("form", form)
		return modelAndView
	}

}